libraries/AP_EFI/AP_EFI_Serial_Lutan.cpp

```diff
@@ -0,0 +1,91 @@
+/*
+   This program is free software: you can redistribute it and/or modify
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation, either version 3 of the License, or
+   (at your option) any later version.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+ 
+#include <AP_HAL/AP_HAL.h>
+#include "AP_EFI_Serial_Lutan.h"
+#include <AP_HAL/utility/sparse-endian.h>
+#include <stdio.h>
+
+#if HAL_EFI_ENABLED
+#include <AP_SerialManager/AP_SerialManager.h>
+
+extern const AP_HAL::HAL &hal;
+
+AP_EFI_Serial_Lutan::AP_EFI_Serial_Lutan(AP_EFI &_frontend):
+    AP_EFI_Backend(_frontend)
+{
+    port = AP::serialmanager().find_serial(AP_SerialManager::SerialProtocol_EFI, 0);
+}
+
+
+void AP_EFI_Serial_Lutan::update()
+{
+    if (!port) {
+        return;
+    }
+
+    uint32_t now = AP_HAL::millis();
+    if (pkt_nbytes > 0 && now - last_recv_ms > 100) {
+        pkt_nbytes = 0;
+    }
+    const uint32_t n = MIN(sizeof(pkt), port->available());
+    if (n > 0) {
+        last_recv_ms = now;
+    }
+    if (n + pkt_nbytes > sizeof(pkt)) {
+        pkt_nbytes = 0;
+    }
+    pkt_nbytes += port->read(&pkt[pkt_nbytes], n);
```

Comment:
`read()` can return a negative value.  That could result in a bad memory access.

Reply:
only if available() is broken. If it is then we have bigger problems

Reply:
done